### PR TITLE
Derandomize std.algorithm

### PR DESCRIPTION
**Pull request by wilzbach**
So it turns out that using a random seed leads to "interesting" effects. One example is that our coverage report jumps _randomly_ up & down.

Don't believe me? Check it out for yourself:

https://codecov.io/gh/dlang/phobos/compare/77f1a9067b59395f8444cb2d01047680aebcfe5f...f2b58341725a6d195b388c1b6f31dcdaa9f000f7/changes
https://codecov.io/gh/dlang/phobos/compare/22331db571d874ba44e412777ed8310536a18dff...89e4014e3d89c7de4d2e2c50eca7924e465ff062/changes
https://codecov.io/gh/dlang/phobos/compare/86050a1c59eda327fa263549f0b53e56a4cfa9ed...2749175ceb24e4b780725677b7e233f65e4f791c/changes
https://codecov.io/gh/dlang/phobos/compare/77f1a9067b59395f8444cb2d01047680aebcfe5f...2e05f29d3c4cc7690198c3ad0d06eb23293da48a/changes
...
(endless list)

I think it's quite reasonable to fix the random test to a constant value, s.t. we know what parts are actually covered and if needed add tests for the uncovered bits.

Note that there's also floppy line in `std.parallelism`, but that one might be harder to get.
edit: for `std.parallelism` we should probably have a look at this https://github.com/dlang/phobos/pull/4399

### CODE REVIEW

**Comment by andralex** [comment]
Hmmm... there's good advantage to randomized unittesting, we don't want to give that up. How about non-random tests that ensure coverage and then a randomized unittest? Would that be difficult?

**Comment by wilzbach** [comment]
> How about non-random tests that ensure coverage and then a randomized unittest? Would that be difficult?

So here's what I did:
- run coverage and save as `before.lst`
- comment all random tests in `sorting.d`
- run coverage and save as `after.lst`
- LOOP:
- diff both `.lst` files and search for uncovered bits
- add `assert(0)` in the uncovered bit to find the regarding test
- print out the seed and add to the list of pre-defined seeds
- continue LOOP until no coverage difference is found

> 88.629% (+0.007%) compared to 0fef09a

Seems like it worked :)

**Comment by schveiguy** [comment]
From a practical position, I don't like random unit tests, because there can be weird interactions (e.g. some PR fails for totally unrelated random unit test failure, cannot repeat it).

However, I agree random testing has good benefits. I would propose that we have true random testing, but not part of unit tests. Make it a separate testing system, such that random failures generate alerts (probably posts to mailing list), with enough info to add the failures as non-random tests. But they aren't creating spurious failures for PR testing. Something to think about, probably not a quick solution.

In a previous life I wrote problem sets for topcoder and also tested other's problem sets. One thing I would do frequently is generate tons of random data to see if I could cause my solution or the other solutions to fail. If I found a test case that caused failure, then I would add it as a non-random test case for the testing phase of the competition (the hypothesis being that if one of our solutions got caught by some corner case exposed by the random data, someone else might too). During the actual competition, all solutions were run against the exact same test cases to judge whether they were correct or not. So the results are objective, even if they might have been incomplete. Such a stance is well founded.

**Comment by andralex** [comment]
Yah, printing the seed along with indications on how to set it to repro is essential. BTW at best we should have a systematic approach (one global seed, one easy way to set etc) instead of each unittest doing its own.

**Comment by wilzbach** [comment]
> From a practical position, I don't like random unit tests, because there can be weird interactions (e.g. some PR fails for totally unrelated random unit test failure, cannot repeat it).

You can scratch the "can be weird interactions" - just have a look at the CodeCov CI project status of a recent PR - it's moving "randomly" ;-)

> While we are at it, could you please change them to print the seed when the tests fail for reproducibility? 
> Yah, printing the seed along with indications on how to set it to repro is essential. BTW at best we should have a systematic approach (one global seed, one easy way to set etc) instead of each unittest doing its own.

Could we please go step by step?
While having a common seeding mechanism in Phobos would be nice, (1) I would prefer not to have them in the first place, (2) the other randomized tests seem to test not such complicated algorithms, (3) it's a design decision decision (e.g. `mixin failsafeSeed`, `runInSeededEnv`, `shared seed`, ...) that blocks this PR unnecessarily, (4) my interest is atm primarily in fixing the random "red cross" for PRs rather soon than later and this PR is hopefully a first step towards it.

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -883,26 +883,31 @@ if (ss == SwapStrategy.unstable && isRandomAccessRange!Range
...
 
-    auto a = new int[](uniform(0, 100));
-    foreach (ref e; a)
-    {
-        e = uniform(0, 50);
-    }
-    auto pieces = partition3(a, 25);
-    assert(pieces[0].length + pieces[1].length + pieces[2].length == a.length);
-    foreach (e; pieces[0])
+    uint[] seeds = [3923355730, 1927035882, unpredictableSeed];
```
Comment (klickverbot) [inline std/algorithm/sorting.d]: Stylistic suggestion: `immutable seeds = …`

File: std/algorithm/sorting.d [hunk 2]
```diff
@@ -3500,24 +3505,30 @@ private T[] randomArray(Flag!"exactSize" flag = No.exactSize, T = int)(
...
-        foreach (ref e; a) e = uniform(-1000, 1000);
-    auto k = uniform(0, a.length);
-    topN(a, k);
-    if (k > 0)
-    {
-        auto left = reduce!max(a[0 .. k]);
-        assert(left <= a[k]);
-    }
-    if (k + 1 < a.length)
+    uint[] seeds = [90027751, 2709791795, 1374631933, 995751648, 3541495258, 984840953, unpredictableSeed];
```
Comment (klickverbot) [inline std/algorithm/sorting.d]: Same here: `const`/`immutable`

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -3500,24 +3505,30 @@ private T[] randomArray(Flag!"exactSize" flag = No.exactSize, T = int)(
...
+    foreach (s; seeds)
     {
-        auto right = reduce!min(a[k + 1 .. $]);
-        assert(right >= a[k]);
+        auto r = Random(s);
+
+        int[] a = new int[uniform(1, 10000, r)];
+            foreach (ref e; a) e = uniform(-1000, 1000, r);
```
Comment (klickverbot) [inline std/algorithm/sorting.d]: Wonky indentation.

**Comment by schveiguy** [comment]
> BTW at best we should have a systematic approach (one global seed, one easy way to set etc) instead of each unittest doing its own.

Careful here, I think if you are having a random test, you need the seed printed before each test, not one global seed. Otherwise, other factors (which modules are included, which ones test random data, etc) can affect whether the seed is set correctly at the time of the test. If you just mean there needs to be a standardized way to set the seed for an individual test, yes, that is ideal.

Yes, fix the issue that is causing random PR failures first, and then maybe we can have a bigger discussion on random testing elsewhere.

File: std/algorithm/sorting.d (hunk 3 above)
Comment (wilzbach) [inline std/algorithm/sorting.d]: That's from the existing source ;-)

File: std/algorithm/sorting.d (hunk 2 above)
Comment (wilzbach) [inline std/algorithm/sorting.d]: `Random` or in particular the `MersenneTwisterEngine` doesn't seem to be so convinced of the concepts of immutable seeds:

```
std/algorithm/sorting.d(891): Error: constructor std.random.MersenneTwisterEngine!(uint, 32LU, 624LU, 397LU, 31LU, 2567483615u, 11LU, 4294967295u, 7LU, 2636928640u, 15LU, 4022730752u, 18LU, 1812433253u).MersenneTwisterEngine.this (uint value) is not callable using argument types (immutable(long))
```

**Comment by wilzbach** [comment]
> Fair enough – lgtm, feel free to merge after fixing the broken indentation.
..
> Yes, fix the issue that is causing random PR failures first, 

Done, but a PR needs to be approved via an GH review otherwise merging is blocked.
If you want, you could give the new "auto-merge-squash" label a try ;-)

**Comment by wilzbach** [comment]
> I didn't realise one couldn't green-light their own PR's.

This is a rather new change due to us trying to catching up with the improvements GH makes ;-)
If case you haven't seen this summary, it might be interesting as well:

http://forum.dlang.org/post/qoamackqqnkqrxkochdu@forum.dlang.org
(or summarized as Wiki entry: https://wiki.dlang.org/Guidelines_for_maintainers)

**Comment by andralex** [comment]
Thanks @wilzbach this is terrific work.

**Comment by andralex** [comment]
@schveiguy well unless there's randomization in the order of running unittests, we can set things up so one seed sets up an entire test battery. At any rate please let's not print random numbers with every successful run :).

Agreed on taking the discussion to a higher level, maybe you could lead it. Thanks!

**Comment by schveiguy** [comment]
> At any rate please let's not print random numbers with every successful run

Right, I was thinking only printing after a failure, but I meant you need to look at the seed for the specific test (and store it before running the test), not once for the beginning of all unit tests.

**Comment by wilzbach** [comment]
> Huh?

image

@andralex could you please have a look? (and maybe for the other repos as well)

**Comment by wilzbach** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
> Auto-merge toggled on

The labelled PRs [still don't have priority on the auto-tester](https://github.com/dlang-bots/dlang-bot/pull/50). Hence I went for the faster option and squashed manually.

**Comment by andralex** [comment]
Alright, squash merges should be on now. Thanks for your diligence!
